### ISSUE TITLE
Rebalacing `macro_metavar_expr` to allow stabilization

### ISSUE THREAD

**Issue by c410-f3r** [rust-lang/compiler-team#680]
Too small to be a RFC and too complicated to be a PR. I guess this is the best place but if not, feel free to suggest another location.

# Proposal

`macro_metavar_expr` is an ***approved*** feature that unfortunately created divergent opinions ***after*** its acceptance, which ended up blocking stabilization.

In hopes of making some progress, two approaches based on the feedback provided so far will be proposed here.

## 1. Innermost vs Outermost indexes

`count` uses outermost indices while `length` uses innermost indices and this inconsistency creates unnecessary confusion.

![meta](https://user-images.githubusercontent.com/17877264/158820060-035b2548-9ea9-4cbb-850b-e5836cf892fe.png)

To improve the situation, the order of all elements should start from the innermost index to the outermost index.

### Mentions

* https://github.com/rust-lang/rust/issues/83527#issuecomment-1068046662
* https://github.com/rust-lang/rust/issues/83527#issuecomment-1068525894
* https://github.com/rust-lang/rust/issues/83527#issuecomment-1069173636
* https://github.com/rust-lang/rust/issues/83527#issuecomment-1069635121
* https://github.com/rust-lang/rust/issues/83527#issuecomment-1070308699
* https://github.com/rust-lang/rust/issues/83527#issuecomment-1070366701

## 2. `$` prefix

Taking `count` as an example, should the syntax be `count(some_metavariable)` or `count($some_metavariable)`? The original RFC specified that metavariable expressions should refer metavariables without `$` prefixes but there were some arguments in favour of `$`.

For unblocking purposes, the requirement of `$` is being proposed. Such enforcement doesn't appear to incur a significant overhead besides the additional typing and interactions with `$$` or multiple `$$`s shouldn't be a problem as long as the final expanded `$` refers a metavariable.

### Mentions

* https://github.com/rust-lang/rust/pull/95860#issuecomment-1125684148
* https://github.com/rust-lang/rust/pull/95860#issuecomment-1126403309
* https://github.com/rust-lang/rust/pull/95860#issuecomment-1126487232
* https://github.com/rust-lang/rust/pull/95860#issuecomment-1139881538
* https://internals.rust-lang.org/t/macro-meta-functions/16743
* https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/macro.20metafunctions.20vs.20eager.20expansion

# Mentors or Reviewers

@c410-f3r

# Process

The main points of the [Major Change Process][MCP] are as follows:

* [x] File an issue describing the proposal.
* [ ] A compiler team member or contributor who is knowledgeable in the area can **second** by writing `@rustbot second`.
    * Finding a "second" suffices for internal changes. If however, you are proposing a new public-facing feature, such as a `-C flag`, then full team check-off is required.
    * Compiler team members can initiate a check-off via `@rfcbot fcp merge` on either the MCP or the PR.
* [ ] Once an MCP is seconded, the Final Comment Period begins. If no objections are raised after 10 days, the MCP is considered **approved**.

**Comment by Nilstrieb** [comment]
I don't think a compiler MCP is the right place for this. I think you're right that it's not an RFC either. Perhaps just an IRLO thread or a lang zulip thread or on a separate issue?

**Comment by c410-f3r** [comment]
Closing per the above comment. I will then try in the tracking issue.
